src/libmongoc/tests/unified/entity-map.c

```diff
@@ -0,0 +1,603 @@
+   if (!session_opts) {
+      goto done;
+   }
+   entity->value = mongoc_client_start_session (client, session_opts, error);
+   if (!entity->value) {
+      goto done;
+   }
+   ret = true;
+done:
+   mongoc_session_opts_destroy (session_opts);
+   bson_free (client_id);
+   bson_destroy (session_opts_bson);
+   bson_parser_destroy (parser);
+   if (!ret) {
+      entity_destroy (entity);
+      return NULL;
+   }
+   return entity;
+}
+
+/* Caveat: The spec encourages, but does not require, that entities are defined
+ * in dependency order:
+ * "Test files SHOULD define entities in dependency order, such that all
+ * referenced entities (e.g. client) are defined before any of their dependent
+ * entities (e.g. database, session)."
+ * If a test ever does break this pattern (flipping dependency order), that can
+ * be solved by creating C objects lazily in entity_map_get.
+ * The current implementation here does the simple thing and creates the C
+ * object immediately.
+ */
+bool
+entity_map_create (entity_map_t *entity_map, bson_t *bson, bson_error_t *error)
+{
+   bson_iter_t iter;
+   const char *entity_type;
+   bson_t entity_bson;
+   entity_t *entity = NULL;
+   entity_t *entity_iter;
+   bool ret = false;
+
+   bson_iter_init (&iter, bson);
+   if (!bson_iter_next (&iter)) {
+      test_set_error (error, "Empty entity");
+      goto done;
+   }
+
+   entity_type = bson_iter_key (&iter);
+   bson_iter_bson (&iter, &entity_bson);
+   if (bson_iter_next (&iter)) {
+      test_set_error (error,
+                      "Extra field in entity: %s: %s",
+                      bson_iter_key (&iter),
+                      tmp_json (bson));
+      goto done;
+   }
+
+   if (0 == strcmp (entity_type, "client")) {
+      entity = entity_client_new (&entity_bson, error);
+   } else if (0 == strcmp (entity_type, "database")) {
+      entity = entity_database_new (entity_map, &entity_bson, error);
+   } else if (0 == strcmp (entity_type, "collection")) {
+      entity = entity_collection_new (entity_map, &entity_bson, error);
+   } else if (0 == strcmp (entity_type, "session")) {
+      entity = entity_session_new (entity_map, &entity_bson, error);
+   } else {
+      test_set_error (
+         error, "Unknown entity type: %s: %s", entity_type, tmp_json (bson));
+      goto done;
+   }
+
+   if (!entity) {
+      goto done;
+   }
+
+   LL_FOREACH (entity_map->entities, entity_iter)
+   {
+      if (0 == strcmp (entity_iter->id, entity->id)) {
+         test_set_error (
+            error, "Attempting to create duplicate entity: '%s'", entity->id);
+         entity_destroy (entity);
+         goto done;
+      }
+   }
+
+   ret = true;
+done:
+   if (!ret) {
+      entity_destroy (entity);
+   } else {
+      LL_PREPEND (entity_map->entities, entity);
+   }
+   return ret;
+}
+
+static void
+entity_destroy (entity_t *entity)
+{
+   event_t *event;
+   event_t *tmp;
+
+   if (!entity) {
+      return;
+   }
+   if (entity->type) {
```

Comment:
when would we expect to have an entity that didn't have a type?

Reply:
Good question. Entities should always have a type. I changed this to an assertion.